File: src/A_mul_B.jl

export A_mul_B!

function A_mul_B!( alpha::Float64,
                   A::SparseMatrixCSC{Float64,Ti},
                   x::Array{Float64},
                   beta::Float64,
                   y::Array{Float64},
                   nthreads::Int64=0 ) where Ti
# Real:  y = beta*y  +  alpha * A*x 

   if nthreads == 0
      #Base.A_mul_B!( alpha, A, x, beta, y )
	  mul!(y,A,x, alpha, beta)
      return
   elseif nthreads < 1
      throw(ArgumentError("nthreads < 1"))
   end

	n,m = size(A)
    nvec = size(x,2)

   if size(x,1) != m || size(y,1) != n 
      throw(DimensionMismatch("length(x) != m || length(y) != n"))
   elseif size(y,2) != nvec
      throw(DimensionMismatch("length(y,2) != nvec"))
   end
   
   
	if Ti == Int32
      p = ccall( (:a_mul_b_rr_32_, spmatveclib),
		       Nothing, ( Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Float64}, Ref{Float64}, Ref{Float64}, Ref{Int32}, Ref{Int32}, Ref{Float64}, Ref{Float64}),
             nthreads, nvec, m, n,    alpha,   beta,              A.nzval,      A.rowval,   A.colptr,   x,   y)
   elseif Ti == Int64
      p = ccall( (:a_mul_b_rr_64_, spmatveclib),
		       Nothing, ( Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Float64}, Ref{Float64}, Ref{Float64}, Ref{Int64}, Ref{Int64}, Ref{Float64}, Ref{Float64}),
             nthreads, nvec, m, n,    alpha,   beta,              A.nzval,      A.rowval,   A.colptr,   x,   y) 
   else
      error("Unsupported sparse matrix indexing integer type $Ti")
   end
   return p
end  # function A_mul_B!

#------------------------------------------------------------------------------

function A_mul_B!( alpha::ComplexF64,
                   A::SparseMatrixCSC{Float64,Ti},
                   x::Array{ComplexF64},
                   beta::ComplexF64,
                   y::Array{ComplexF64},
                   nthreads::Int64=0 ) where Ti
# Real, Complex A:  y = beta*y  +  alpha * A*x 

   if nthreads == 0
      mul!(y,A,x, alpha, beta) # Base.A_mul_B!( alpha, complex(A), x, beta, y )
      return
   elseif nthreads < 1
      throw(ArgumentError("nthreads < 1"))
   end

	n,m  = size(A)
   nvec = size(x,2)

   if size(x,1) != m || size(y,1) != n 
      throw(DimensionMismatch("length(x) != m || length(y) != n"))
   elseif size(y,2) != nvec
      throw(DimensionMismatch("length(y,2) != nvec"))
   end
   
	if Ti == Int32
	   p = ccall( (:a_mul_b_rc_32_, spmatveclib),
		 Nothing, ( Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{ComplexF64}, Ref{ComplexF64}, Ref{Float64}, Ref{Int32}, Ref{Int32}, Ref{ComplexF64}, Ref{ComplexF64}),
                   nthreads, nvec, m, n,     alpha,   beta,              A.nzval,      A.rowval,   A.colptr,   x,  y)
   elseif Ti == Int64
      p = ccall( (:a_mul_b_rc_64_, spmatveclib),
		 Nothing, ( Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{ComplexF64}, Ref{ComplexF64}, Ref{Float64}, Ref{Int64}, Ref{Int64}, Ref{ComplexF64}, Ref{ComplexF64}),
                   nthreads, nvec, m, n,     alpha,   beta,              A.nzval,      A.rowval,   A.colptr,   x,  y)
   else
      error("Unsupported sparse matrix indexing integer type $Ti")
   end
   return p
end  # function A_mul_B!

#------------------------------------------------------------------------------

function A_mul_B!( alpha::ComplexF64,
                   A::SparseMatrixCSC{ComplexF64,Ti},
                   x::Array{ComplexF64},
                   beta::ComplexF64,
                   y::Array{ComplexF64},
                   nthreads::Int64=0 ) where Ti
# Complex A:  y = beta*y  +  alpha * A*x 

   if nthreads == 0
      mul!(y,A,x, alpha, beta) # Base.A_mul_B!( alpha, A, x, beta, y )
      return
   elseif nthreads < 1
      throw(ArgumentError("nthreads < 1"))
   end

	n,m  = size(A)
   nvec = size(x,2)

   if size(x,1) != m || size(y,1) != n 
      throw(DimensionMismatch("length(x) != m || length(y) != n"))
   elseif size(y,2) != nvec
      throw(DimensionMismatch("length(y,2) != nvec"))
   end
   
	if Ti == Int32
	   p = ccall( (:a_mul_b_cc_32_, spmatveclib),
		   Nothing, ( Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{ComplexF64}, Ref{ComplexF64}, Ref{ComplexF64}, Ref{Int32}, Ref{Int32}, Ref{ComplexF64}, Ref{ComplexF64}),
                   nthreads, nvec, m, n,     alpha,   beta,            A.nzval,      A.rowval,   A.colptr,  x,  y)
   elseif Ti == Int64
      p = ccall( (:a_mul_b_cc_64_, spmatveclib),
		 Nothing, ( Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{Int64}, Ref{ComplexF64}, Ref{ComplexF64}, Ref{ComplexF64}, Ref{Int64}, Ref{Int64}, Ref{ComplexF64}, Ref{ComplexF64}),
                   nthreads, nvec, m, n,     alpha,   beta,            A.nzval,      A.rowval,   A.colptr,  x,  y)
   else
      error("Unsupported sparse matrix indexing integer type $Ti")
   end
   return p
end  # function A_mul_B!

File: deps/src/A_mul_B.f90

subroutine a_mul_b_rr_32( nthreads, nvec, n, m, alpha, beta, A, jA, iA, x, y )
!DIR$ ATTRIBUTES DLLEXPORT :: a_mul_b_rr_32
!DIR$ ATTRIBUTES ALIAS: 'a_mul_b_rr_32_':: a_mul_b_rr_32

! y = beta*y  +  alpha * A*x

use omp_lib
implicit none

integer(kind=8),intent(in):: nthreads
integer(kind=8),intent(in):: nvec
integer(kind=8),intent(in):: n  ! # of columns in A
integer(kind=8),intent(in):: m  ! # of rows in A

real(kind=8),intent(in):: alpha, beta
real(kind=8),intent(in):: A(*)
integer(kind=4),intent(in):: jA(*), iA(n+1)
real(kind=8),intent(in):: x(n,nvec)
real(kind=8),intent(inout):: y(m,nvec)

integer ivec, i, j1,j2, j, jaj, mythread, mm, jm
real(kind=8) xi
real(kind=8),allocatable:: yt(:)

include "A_mul_B.fi"

return
end subroutine a_mul_b_rr_32

subroutine a_mul_b_rr_64( nthreads, nvec, n, m, alpha, beta, A, jA, iA, x, y )
!DIR$ ATTRIBUTES DLLEXPORT :: a_mul_b_rr_64
!DIR$ ATTRIBUTES ALIAS: 'a_mul_b_rr_64_':: a_mul_b_rr_64

! y = beta*y  +  alpha * A*x

use omp_lib
implicit none

integer(kind=8),intent(in):: nthreads
integer(kind=8),intent(in):: nvec
integer(kind=8),intent(in):: n  ! # of columns in A
integer(kind=8),intent(in):: m  ! # of rows in A

real(kind=8),intent(in):: alpha, beta
real(kind=8),intent(in):: A(*)
integer(kind=8),intent(in):: jA(*), iA(n+1)
real(kind=8),intent(in):: x(n,nvec)
real(kind=8),intent(inout):: y(m,nvec)

integer ivec, i, j1,j2, j, jaj, mythread, mm, jm
real(kind=8) xi
real(kind=8),allocatable:: yt(:)

include "A_mul_B.fi"

return
end subroutine a_mul_b_rr_64

!--------------------------------------------------------------------

subroutine a_mul_b_rc_32( nthreads, nvec, n, m, alpha, beta, A, jA, iA, x, y )
!DIR$ ATTRIBUTES DLLEXPORT :: a_mul_b_rc_32
!DIR$ ATTRIBUTES ALIAS: 'a_mul_b_rc_32_':: a_mul_b_rc_32

! y = beta*y  +  alpha * A*x

use omp_lib
implicit none

integer(kind=8),intent(in):: nthreads
integer(kind=8),intent(in):: nvec
integer(kind=8),intent(in):: n  ! # of columns in A
integer(kind=8),intent(in):: m  ! # of rows in A

complex(kind=8),intent(in):: alpha, beta
real(kind=8),intent(in):: A(*)
integer(kind=4),intent(in):: jA(*), iA(n+1)
complex(kind=8),intent(in):: x(n,nvec)
complex(kind=8),intent(inout):: y(m,nvec)

integer ivec, i, j1,j2, j, jaj, mythread, mm, jm
complex(kind=8) xi
complex(kind=8),allocatable:: yt(:)

include "A_mul_B.fi"

return
end subroutine a_mul_b_rc_32

subroutine a_mul_b_rc_64( nthreads, nvec, n, m, alpha, beta, A, jA, iA, x, y )
!DIR$ ATTRIBUTES DLLEXPORT :: a_mul_b_rc_64
!DIR$ ATTRIBUTES ALIAS: 'a_mul_b_rc_64_':: a_mul_b_rc_64

! y = beta*y  +  alpha * A*x

use omp_lib
implicit none

integer(kind=8),intent(in):: nthreads
integer(kind=8),intent(in):: nvec
integer(kind=8),intent(in):: n  ! # of columns in A
integer(kind=8),intent(in):: m  ! # of rows in A

complex(kind=8),intent(in):: alpha, beta
real(kind=8),intent(in):: A(*)
integer(kind=8),intent(in):: jA(*), iA(n+1)
complex(kind=8),intent(in):: x(n,nvec)
complex(kind=8),intent(inout):: y(m,nvec)

integer ivec, i, j1,j2, j, jaj, mythread, mm, jm
complex(kind=8) xi
complex(kind=8),allocatable:: yt(:)

include "A_mul_B.fi"

return
end subroutine a_mul_b_rc_64

!--------------------------------------------------------------------

subroutine a_mul_b_cc_32( nthreads, nvec, n, m, alpha, beta, A, jA, iA, x, y )
!DIR$ ATTRIBUTES DLLEXPORT :: a_mul_b_cc_32
!DIR$ ATTRIBUTES ALIAS: 'a_mul_b_cc_32_':: a_mul_b_cc_32

! y = beta*y  +  alpha * A*x

use omp_lib
implicit none

integer(kind=8),intent(in):: nthreads
integer(kind=8),intent(in):: nvec
integer(kind=8),intent(in):: n  ! # of columns in A
integer(kind=8),intent(in):: m  ! # of rows in A

complex(kind=8),intent(in):: alpha, beta
complex(kind=8),intent(in):: A(*)
integer(kind=4),intent(in):: jA(*), iA(n+1)
complex(kind=8),intent(in):: x(n,nvec)
complex(kind=8),intent(inout):: y(m,nvec)

integer ivec, i, j1,j2, j, jaj, mythread, mm, jm
complex(kind=8) xi
complex(kind=8),allocatable:: yt(:)

include "A_mul_B.fi"

return
end subroutine a_mul_b_cc_32

subroutine a_mul_b_cc_64( nthreads, nvec, n, m, alpha, beta, A, jA, iA, x, y )
!DIR$ ATTRIBUTES DLLEXPORT :: a_mul_b_cc_64
!DIR$ ATTRIBUTES ALIAS: 'a_mul_b_cc_64_':: a_mul_b_cc_64

! y = beta*y  +  alpha * A*x

use omp_lib
implicit none

integer(kind=8),intent(in):: nthreads
integer(kind=8),intent(in):: nvec
integer(kind=8),intent(in):: n  ! # of columns in A
integer(kind=8),intent(in):: m  ! # of rows in A

complex(kind=8),intent(in):: alpha, beta
complex(kind=8),intent(in):: A(*)
integer(kind=8),intent(in):: jA(*), iA(n+1)
complex(kind=8),intent(in):: x(n,nvec)
complex(kind=8),intent(inout):: y(m,nvec)

integer ivec, i, j1,j2, j, jaj, mythread, mm, jm
complex(kind=8) xi
complex(kind=8),allocatable:: yt(:)

include "A_mul_B.fi"

return
end subroutine a_mul_b_cc_64
